File: backend2/app/routers/fte.py
from fastapi import APIRouter, Depends, HTTPException, Query
from sqlalchemy.orm import Session
from typing import List, Optional
from datetime import date

from app.database import get_db
from app.schemas.fte import (
    FTEScale,
    FTEScaleCreate,
    FTEAssignment,
    FTEAssignmentCreate,
    FTEPagination
)
from app.services.fte_crud import (
    get_fte_scales,
    get_fte_scale,
    create_fte_scale,
    update_fte_scale,
    delete_fte_scale,
    get_fte_assignments,
    get_fte_assignment,
    create_fte_assignment,
    update_fte_assignment,
    delete_fte_assignment
)

router = APIRouter(
    prefix="/api/fte",
    tags=["fte"]
)

# FTE Scale endpoints
@router.get("/scales", response_model=List[FTEScale])
def read_fte_scales(skip: int = 0, limit: int = 100, db: Session = Depends(get_db)):
    scales = get_fte_scales(db, skip=skip, limit=limit)
    return scales

@router.post("/scales", response_model=FTEScale)
def create_fte_scale_route(scale: FTEScaleCreate, db: Session = Depends(get_db)):
    return create_fte_scale(db=db, scale=scale)

@router.put("/scales/{scale_id}", response_model=FTEScale)
def update_fte_scale_route(scale_id: int, scale: FTEScaleCreate, db: Session = Depends(get_db)):
    db_scale = update_fte_scale(db=db, scale_id=scale_id, scale=scale)
    if db_scale is None:
        raise HTTPException(status_code=404, detail="FTE scale not found")
    return db_scale

@router.delete("/scales/{scale_id}")
def delete_fte_scale_route(scale_id: int, db: Session = Depends(get_db)):
    success = delete_fte_scale(db=db, scale_id=scale_id)
    if not success:
        raise HTTPException(status_code=404, detail="FTE scale not found")
    return {"message": "FTE scale deleted successfully"}

# FTE Assignment endpoints
@router.get("/assignments", response_model=FTEPagination)
def read_fte_assignments(
    skip: int = Query(0, ge=0),
    limit: int = Query(10, ge=1, le=100),
    staff_id: Optional[int] = None,
    start_date: Optional[date] = None,
    end_date: Optional[date] = None,
    db: Session = Depends(get_db)
):
    assignments, total = get_fte_assignments(
        db,
        skip=skip,
        limit=limit,
        staff_id=staff_id,
        start_date=start_date,
        end_date=end_date
    )
    return FTEPagination(
        items=assignments,
        total=total,
        page=skip // limit + 1,
        size=limit,
        pages=(total + limit - 1) // limit
    )

@router.post("/assignments", response_model=FTEAssignment)
def create_fte_assignment_route(assignment: FTEAssignmentCreate, db: Session = Depends(get_db)):
    return create_fte_assignment(db=db, assignment=assignment)

@router.put("/assignments/{assignment_id}", response_model=FTEAssignment)
def update_fte_assignment_route(
    assignment_id: int,
    assignment: FTEAssignmentCreate,
    db: Session = Depends(get_db)
):
    db_assignment = update_fte_assignment(db=db, assignment_id=assignment_id, assignment=assignment)
    if db_assignment is None:
        raise HTTPException(status_code=404, detail="FTE assignment not found")
    return db_assignment

@router.delete("/assignments/{assignment_id}")
def delete_fte_assignment_route(assignment_id: int, db: Session = Depends(get_db)):
    success = delete_fte_assignment(db=db, assignment_id=assignment_id)
    if not success:
        raise HTTPException(status_code=404, detail="FTE assignment not found")
    return {"message": "FTE assignment deleted successfully"} 

File: frontend/src/services/fte.ts
import axios from 'axios'

const API_BASE_URL = import.meta.env.VITE_API_BASE_URL || 'http://localhost:8001'

export interface FTEScale {
  id: string
  name: string
  type: "shifts" | "hours" | "days"
  baseValue: number
  equivalentFTE: number
}

export interface FTEAssignment {
  id: string
  scaleId: string
  targetType: "role" | "staff" | "group"
  targetId: string
  value: number
}

export interface FTEPagination {
  items: FTEAssignment[]
  total: number
  page: number
  size: number
  pages: number
}

export interface CreateFTEScale {
  name: string
  type: "shifts" | "hours" | "days"
  baseValue: number
  equivalentFTE: number
}

export interface CreateFTEAssignment {
  scaleId: string
  targetType: "role" | "staff" | "group"
  targetId: string
  value: number
}

export const fteService = {
  // FTE Scale endpoints
  async getFTEScales(): Promise<FTEScale[]> {
    try {
      const response = await axios.get(`${API_BASE_URL}/api/fte/scales`)
      return response.data || []
    } catch (error) {
      console.error('Error fetching FTE scales:', error)
      return []
    }
  },

  async createFTEScale(data: CreateFTEScale): Promise<FTEScale> {
    try {
      const response = await axios.post(`${API_BASE_URL}/api/fte/scales`, data)
      return response.data
    } catch (error) {
      console.error('Error creating FTE scale:', error)
      throw error
    }
  },

  async updateFTEScale(id: string, data: CreateFTEScale): Promise<FTEScale> {
    try {
      const response = await axios.put(`${API_BASE_URL}/api/fte/scales/${id}`, data)
      return response.data
    } catch (error) {
      console.error('Error updating FTE scale:', error)
      throw error
    }
  },

  async deleteFTEScale(id: string): Promise<void> {
    try {
      await axios.delete(`${API_BASE_URL}/api/fte/scales/${id}`)
    } catch (error) {
      console.error('Error deleting FTE scale:', error)
      throw error
    }
  },

  // FTE Assignment endpoints
  async getFTEAssignments(page: number = 1, size: number = 10): Promise<FTEPagination> {
    try {
      const response = await axios.get(`${API_BASE_URL}/api/fte/assignments`, {
        params: { page, size }
      })
      return response.data || { items: [], total: 0, page, size, pages: 0 }
    } catch (error) {
      console.error('Error fetching FTE assignments:', error)
      return { items: [], total: 0, page, size, pages: 0 }
    }
  },

  async createFTEAssignment(data: CreateFTEAssignment): Promise<FTEAssignment> {
    try {
      const response = await axios.post(`${API_BASE_URL}/api/fte/assignments`, data)
      return response.data
    } catch (error) {
      console.error('Error creating FTE assignment:', error)
      throw error
    }
  },

  async updateFTEAssignment(id: string, data: CreateFTEAssignment): Promise<FTEAssignment> {
    try {
      const response = await axios.put(`${API_BASE_URL}/api/fte/assignments/${id}`, data)
      return response.data
    } catch (error) {
      console.error('Error updating FTE assignment:', error)
      throw error
    }
  },

  async deleteFTEAssignment(id: string): Promise<void> {
    try {
      await axios.delete(`${API_BASE_URL}/api/fte/assignments/${id}`)
    } catch (error) {
      console.error('Error deleting FTE assignment:', error)
      throw error
    }
  }
} 